notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "For one [jaxopt.implicit_diff.custom root](https://jaxopt.github.io/stable/_autosummary/jaxopt.implicit_diff.custom_root.html#jaxopt.implicit_diff.custom_root) requires a function that takes the parameter that we want to differentiate as an explicit input. To this end we define a new explicit_force_fn as\n",
+    "```python \n",
+    "explicit_force_fn = jit(lambda R, p: force_fn(R,**p))\n",
+    "```\n",
+    "\n",
+    "Furthermore we cannot pass our original force_fn to our solver as this will cause a ```CustomVJPException```. We can just wrap the output of our ```force_fn``` with ```lax.stop_gradient``` in order to fix this issue.\n",
+    "```python\n",
+    "no_grad_force_fn = jit(lambda x: lax.stop_gradient(force_fn(x)))\n",
+    "```\n",
+    "\n",
+    "Having done that we can define our ```solver``` which takes two parameters as its input. A dummy variable ```params```, which we can just delete as we do not need it and a variable ```x``` which is our set of initial positions. It is necessary that we pass our newly defined ```no_grad_force_fn``` to our solver.\n",
+    "\n",
+    "Now we can put it all together and define a ```decorated_solver``` using ```jaxopt``` in order to be able to efficiently differentiate through it:\n",
+    "```python\n",
+    "decorated_solver = implicit_diff.custom_root(explicit_force_fn)(solver)\n",
+    "```\n",
+    "Here we have to use our ```explicit_force_fn```. It is also possible to use a different linear solver for the ```vjp``` computation using the ```solve``` argument of `implicit_diff.custom_root`. Jax comes with a handfull of sparse linear solvers in ```jax.scipy.sparse.linalg```."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "removed-completion",
+   "metadata": {},
+   "source": [
+    "# Adding neighbor lists and other auxiliary output to our solver"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "wound-nirvana",
+   "metadata": {},
+   "source": [
+    "Until now we have only worked with solvers that return one parameter `R_final`. Now we'll see how we have to change our `implicit_diff` function in order to be able to also have our solver return a neighbor list `nbrs` and the number of steps `num_steps` needed to reach the minimium.\n",
+    "\n",
+    "There is only one thing we have to change besides the normal stuff we have to change when using neighbor lists. We simply add the ```has_aux=True``` keyword to ```implicit_diff.custom_root```."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "massive-formation",
+   "metadata": {},
+   "source": [
+    "There's one more funny thing to note about the neighbor list version. Instead of getting an ```CustomVJPException``` when we forget to wrap our ```force_fn``` with ```lax.stop_gradient``` we get no explicit error message but we now get a memory leak instead which drastically slows down the computation. So pay attention that your are passing the correct version of your ```force_fn``` to your ```solver```!"
```

Comment:
Would it be worth filing a bug against JaxOpt? This seems like kind of dangerous behavior.

Reply:
I'll also fill an issue about that with the jaxopt folks.